src/compiler/scala/tools/nsc/settings/PathFactory.scala

```diff
@@ -0,0 +1,27 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.tools.nsc.settings
+
+import scala.reflect.io.{AbstractFile, PlainFile}
+
+/** Converts paths provided in compiler options (e.g elements of `-classpath` or the target directory of `-d`) to
+ *  an `AbstractFile`. */
+trait PathFactory {
+  def getDirectory(path: String): AbstractFile
```

Comment:
Needs documentation about:
  - that `directory` means an existing directory, JAR, or directory-like structure
  - this should return `null` when for physical directories that don't exist.